src/main/java/xyz/nucleoid/plasmid/util/BucketScanner.java

```diff
@@ -0,0 +1,231 @@
+import it.unimi.dsi.fastutil.longs.LongSet;
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.server.world.ServerWorld;
+import net.minecraft.structure.rule.RuleTest;
+import net.minecraft.tag.Tag;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+
+import java.util.*;
+import java.util.function.Predicate;
+
+/**
+ * Methods for finding connected blocks as a {@link LongSet}. Use {@link BlockPos#fromLong} to retrieve these positions.
+ *
+ * @see <a href="https://en.wikipedia.org/wiki/Pixel_connectivity#3-dimensional">3-dimensional pixel connectivity</a>
+ */
+public final class BucketScanner {
+    private BucketScanner() {
+    }
+
+    /**
+     * Finds any 6-connected blocks and puts them in a {@link LongSet}.
+     *
+     * @param origin   the position of the first block
+     * @param depth    the amount of maximum blocks to find
+     * @param ruleTest the rule test for the blocks that can be accepted in the set
+     */
+    public static LongSet findSix(BlockPos origin, int depth, RuleTest ruleTest, ServerWorld world, Random random) {
+        return findSix(origin, depth, pos -> ruleTest.test(world.getBlockState(pos), random));
+    }
+
+    /**
+     * Finds any 6-connected blocks and puts them in a {@link LongSet}.
+     *
+     * @param origin the position of the first block
+     * @param depth  the amount of maximum blocks to find
+     * @param tag    the tag for the blocks that can be accepted in the set
+     */
+    public static LongSet findSix(BlockPos origin, int depth, Tag<Block> tag, ServerWorld world) {
+        return findSix(origin, depth, pos -> world.getBlockState(pos).isIn(tag));
+    }
+
+    /**
+     * Finds any 6-connected blocks and puts them in a {@link LongSet}.
+     *
+     * @param origin the position of the first block
+     * @param depth  the amount of maximum blocks to find
+     * @param block  the block type that can be accepted in the set
+     */
+    public static LongSet findSix(BlockPos origin, int depth, Block block, ServerWorld world) {
+        return findSix(origin, depth, pos -> world.getBlockState(pos).isOf(block));
+    }
+
+    /**
+     * Finds any 6-connected blocks and puts them in a {@link LongSet}.
+     *
+     * @param origin    the position of the first block
+     * @param depth     the amount of maximum blocks to find
+     * @param predicate the predicate for the blocks that can be accepted in the set
+     */
+    public static LongSet findSix(BlockPos origin, int depth, Predicate<BlockPos> predicate) {
+        LongSet set = new LongArraySet();
+        Deque<BlockPos> ends = new ArrayDeque<>();
+        ends.push(origin);
+        BlockPos.Mutable local = origin.mutableCopy();
+        while(depth > 0) {
+            if(ends.isEmpty()) {
+                return set;
+            }
+            BlockPos pos = ends.pollLast();
+            for(Direction direction : Direction.values()) {
```

Comment:
It seems there's a bit of duplication with the logic that performs the search- potentially this type could be an enum for each neighbor handling method, and each implementation implements that neighbor logic specifically?

Reply:
Done in e6cfa44! 😄